### PR TITLE
Scaffold Mamba POC: seam, data pipeline, train/eval skeletons (M1-4)

### PR DESCRIPTION
**Pull request by travisgalloway**
Lay down the full repo structure for the Apple Silicon Mamba POC behind one
hardware seam (ModelInterface), per the milestones 1-4 plan.

Runnable in-container (Linux):
- Portable seam (interface.py, blocks.py config) with import-guard test
- Data pipeline: download/tokenize/pack(uint16)/split/loader, unit-tested
- schedule (warmup+cosine), checkpoint (portable safetensors + within-backend
  resume), val_loss (Tier-1 perplexity) - unit-tested
- 10/10 tests pass; offline data CLI verified end-to-end

Skeletons/stubs (completed + run on a Mac, MLX is Apple-Silicon only):
- mlx_backend (RMSNorm/SelectiveSSM/MambaBlock/model), training loop,
  forward_step_parity, scripts/smoke_test (the M4 gate)

Deferred stubs: cuda_backend + backend_parity, serve/sessions+rewind,
eval/olmes_adapter

Locked: poc d_model 768 / 24 layers / seq 1024 / ~3B tokens (tied embed);
toy d_model 64 / 2 layers / seq 128 / fp32; conformance compared in fp32.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+data = ["datasets>=2.0", "transformers>=4.40"]
+# Apple Silicon backend (NOT installable on Linux/CUDA hosts).
+mlx = ["mlx>=0.18"]
+# Experiment logging.
+logging = ["wandb>=0.16"]
+dev = ["pytest>=8.0"]
+
+[tool.setuptools]
+packages = ["src"]
+
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply pyproject.toml]: ✅ Fixed in 22ec3b5 — switched to setuptools package discovery (include=["src*"]) so subpackages install

File: tests/test_import_guard.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    # Drop any backend modules a prior test may have loaded.
+    for name in list(sys.modules):
+        if name.split(".")[0] in ("mlx", "torch"):
+            del sys.modules[name]
+
+    for mod in PORTABLE_MODULES:
+        importlib.import_module(mod)
+
+    leaked = [m for m in sys.modules if m in FORBIDDEN]
+    assert not leaked, f"backend leaked above the seam: {leaked}"
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply tests/test_import_guard.py]: ✅ Fixed in 22ec3b5 — leak check now matches by top-level package (catches mlx.core, torch._C)

File: src/train/checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    path = str(path)
+    meta = json.loads(Path(path, "resume_meta.json").read_text())
+    meta["optimizer"] = optimizer_deserializer(str(Path(path, "optimizer.state")))
+    return meta
+
+
+def _jsonable(x: Any) -> Any:
+    if isinstance(x, np.ndarray):
+        return {"__ndarray__": x.tolist(), "dtype": str(x.dtype)}
+    return x
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/checkpoint.py:93]: ✅ Fixed in 22ec3b5 — _jsonable is now recursive (nested dicts/lists/np scalars), so RNG states serialize

File: src/train/schedule.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+class CosineSchedule:
+    base_lr: float
+    warmup_steps: int
+    total_steps: int
+    min_lr_ratio: float = 0.1  # floor as a fraction of base_lr; never 0
+
+    def lr_at(self, step: int) -> float:
+        if step < 0:
+            raise ValueError("step must be >= 0")
+        floor = self.base_lr * self.min_lr_ratio
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/schedule.py:36]: ✅ Fixed in 22ec3b5 — added __post_init__ validation (warmup/total/min_lr_ratio)

File: src/train/checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    optimizer_serializer(str(Path(path, "optimizer.state")))
+
+
+def load_resume(path: str, optimizer_deserializer: Callable[[str], Any]) -> dict:
+    """Restore the resume bundle. Returns {step, rng_state, optimizer}."""
+    path = str(path)
+    meta = json.loads(Path(path, "resume_meta.json").read_text())
+    meta["optimizer"] = optimizer_deserializer(str(Path(path, "optimizer.state")))
+    return meta
+
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/checkpoint.py:76]: ✅ Fixed in 22ec3b5 — added tests/test_checkpoint.py (weights roundtrip + resume bundle with a real RNG state)

File: src/eval/val_loss.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    total, n = 0.0, 0
+    for i, (inputs, targets) in enumerate(loader.epoch()):
+        if max_batches is not None and i >= max_batches:
+            break
+        logits = to_numpy(model.forward(inputs))
+        total += cross_entropy(logits, targets)
+        n += 1
+    mean_ce = total / max(1, n)
+    return {"val_loss": mean_ce, "val_perplexity": perplexity(mean_ce)}
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/eval/val_loss.py]: ✅ Fixed in 22ec3b5 — evaluate() now weights cross-entropy by token count

File: src/data/pack.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    """Write a flat uint16 token file. Returns the number of tokens written."""
+    out_path = Path(out_path)
+    out_path.parent.mkdir(parents=True, exist_ok=True)
+
+    if isinstance(ids, np.ndarray):
+        arr = ids.astype(DTYPE, copy=False)
+        if arr.max(initial=0) >= 65536:
+            raise ValueError("token id >= 65536 does not fit uint16")
+        arr.tofile(out_path)
+        n = arr.size
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/data/pack.py:35]: ✅ Fixed in 22ec3b5 — validate ids against [0,65535] before casting to uint16

File: src/train/logging.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+from typing import Optional
+
+
+class JsonlLogger:
+    def __init__(self, path: str, echo: bool = True):
+        self.path = Path(path)
+        self.path.parent.mkdir(parents=True, exist_ok=True)
+        self.echo = echo
+        self._f = self.path.open("w")
+
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/logging.py:21]: ✅ Fixed in 22ec3b5 — docstring aligned with truncate behavior; added append=True for resume

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (travisgalloway) [reply pyproject.toml]: ✅ Fixed in 22ec3b5 — switched to setuptools package discovery (include=["src*"])

File: src/train/checkpoint.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/checkpoint.py:93]: ✅ Fixed in 22ec3b5 — _jsonable is now recursive, so nested RNG states serialize

File: src/train/loop.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  * gradient accumulation
+  * gradient clipping (proven necessary at toy scale)
+  * checkpointing (portable weights + within-backend resume bundle)
+  * logging from step 1: loss, val loss/perplexity, LR, grad norm, tokens/sec
+"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Callable, Optional
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/train/loop.py]: ✅ Fixed in 22ec3b5 — removed unused field import

File: README.md [hunk 10]
```diff
@@ -1,2 +1,68 @@
...
+validated on **Apple Silicon with MLX**, architected behind **one hardware seam**
+so a successful POC migrates to **CUDA** for a larger run with minimal rewrite.
+
+## The seam (most important rule)
+
+All hardware-specific code lives behind `src/model/interface.py`
+(`ModelInterface`). Everything above it — `data/`, `train/`, `serve/`, `eval/`,
+`conformance/` — is portable Python that **never imports MLX or CUDA**. Only
+`src/model/mlx_backend.py` and `src/model/cuda_backend.py` touch a hardware
+library. `tests/test_import_guard.py` enforces this.
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply README.md]: ✅ Fixed in 22ec3b5 — doc updated to include mlx_train_step.py in the seam rule

File: src/model/__init__.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+"""Model package: the hardware seam.
+
+`interface` and `blocks` are portable (no backend imports). `mlx_backend` and
+`cuda_backend` are the only modules permitted to import a hardware library.
+"""
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/model/__init__.py]: ✅ Fixed in 22ec3b5 — doc updated to include mlx_train_step.py in the seam rule

File: docs/MAC_RUNBOOK.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
+# Mac Runbook — Mamba POC next steps
+
+Ordered, local checklist for finishing the POC on Apple Silicon. Mirrors the
+GitHub tracker (parent issue #2) and milestones M1–M4. Each step lists the files
+to touch, the command to run, and the acceptance gate.
+
+> Backend rule: only `src/model/mlx_backend.py` may import `mlx`. Keep everything
+> above the seam backend-free (`tests/test_import_guard.py` enforces this).
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply docs/MAC_RUNBOOK.md]: ✅ Fixed in 22ec3b5 — doc updated to include mlx_train_step.py in the seam rule

File: src/eval/val_loss.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/eval/val_loss.py]: ✅ Fixed in 22ec3b5 — evaluate() now weights cross-entropy by token count
